feat: skip ads where entries cannot be retrieved

bajtos opened the pull request:
While running the indexer in production, I noticed that many index providers return advertisemnt but then they return "404 cid not found" when we ask for the (payload) entries.

Let's recover from that error by ignoring the advertisement and continuing the walk.

Links:
- https://github.com/space-meridian/roadmap/issues/143

bajtos commented on indexer/lib/typings.d.ts:
```diff
@@ -14,41 +14,3 @@ export interface ProviderInfo {
 export type ProviderToInfoMap = Map<string, ProviderInfo>;
 
-/**
-lastAdCID --> [ ] -\
-               ↓    |
-              ...   | entries announced after we started the current walk
-               ↓    |
-              [ ] -/
-               ↓
-     head --> [ ] -\
-               ↓    |
-              ...   | entries visited in this walk
-               ↓    |
-              [ ] -/
-               ↓
-     tail --> [ ] -\
-               ↓    |
-              ...   | entries NOT visited yet
-               ↓    |
-              [ ] -/
-               ↓
- lastHead --> [ ] -\
-               ↓    |
-              ...   | entries visited in the previous walks
-               ↓    |
-              [ ] -/
-               ↓
-             (null)
- */
-export interface WalkerState {
-  head?: string;
-  tail?: string;
-  lastHead?: string;
-  status: string;
-}
-
-export type ProviderToWalkerStateMap = Map<string, WalkerState>
-
-export type PiecePayloadCIDs = string[];
-
```
This is a follow-up for https://github.com/filecoin-station/piece-indexer/pull/18/files#diff-6c702830fadfd5340905905bcfc38b05b06f0c9ebfb92df7afb61e4c4bbeb8eb where I forgot to remove the definition of types re-exported from the shared package.

bajtos replied:
You can find the other definition here:
https://github.com/filecoin-station/piece-indexer/blob/main/repository/lib/typings.d.ts

bajtos commented on indexer/lib/advertisement-walker.js:
```diff
@@ -179,21 +182,26 @@ export async function processNextAdvertisement ({
-      const url = 'url' in err ? err.url : undefined
+      const url = 'url' in err ? err.url : providerInfo.providerAddress
       if ('serverMessage' in err && err.serverMessage) {
         reason = err.serverMessage
+        if ('statusCode' in err && err.statusCode) {
+          reason = `${err.statusCode} ${reason}`
+        }
       } else if ('statusCode' in err && err.statusCode) {
-        reason = `HTTP request to ${url ?? providerInfo.providerAddress} failed: ${err.statusCode}`
+        reason = err.statusCode
       } else if (err.name === 'TimeoutError') {
-        reason = `HTTP request to ${url ?? providerInfo.providerAddress} timed out`
+        reason = 'operation timed out'
       } else if (
         err.name === 'TypeError' &&
         err.message === 'fetch failed' &&
         err.cause &&
         err.cause instanceof Error
       ) {
-        reason = `HTTP request to ${url ?? providerInfo.providerAddress} failed: ${err.cause.message}`
+        reason = err.cause.message
       }
+
+      reason = `HTTP request to ${url} failed: ${reason}`
```
The changes here are not strictly necessary for this fix. If we were running the proposed version, it would have been much easier for me to discover the actual problem.

bajtos commented on indexer/lib/advertisement-walker.js:
```diff
@@ -252,23 +260,43 @@ export async function fetchAdvertisedPayload (providerAddress, advertisementCid,
+        /** @type {any} */(err).serverMessage ?? '<not found>'
+      )
+      return {
+        entriesFetchError: true,
```
This is not the best property name, can you suggest a better one? Or may I should set the value to a string instead of an error?

```suggestion
        entriesFetchError: String(err.statusCode),
```

juliangruber replied:
I was just looking up this property, as was assuming it's one of these
- an `Error`
- an array of entries for which there was a fetch error
- the count of entries for which there was a fetch error

What about `entriesFetchError: err`? Or otherwise `hasEntriesFetchError: true`

juliangruber commented on indexer/test/helpers/http-server.js:
```diff
@@ -0,0 +1,25 @@
+ */
+export async function givenHttpServer (handler) {
+  const server = http.createServer((req, res) => {
+    ;(async () => {
+      await handler(req, res)
+    })().catch(err => {
```
```suggestion
    handler(req, res).catch(err => {
```

bajtos replied:
I accepted your suggestion, but I need to tweak it as TypeScript complains:

```
Property 'catch' does not exist on type 'void'.
```

Solution:

```js
Promise.resolve(handler(req, res)).catch(err => {
```

There is one more subtle difference:
- In my code, if the `handler` function is sync and throws, the error is still handled.
- In your version, such situation triggers "uncaught error"

In that light, I am going to keep my original version.

juliangruber commented on indexer/test/advertisement-walker.test.js:
```diff
@@ -282,7 +278,49 @@ describe('processNextAdvertisement', () => {
+        const frisbeeRes = await fetch(FRISBII_ADDRESS + req.url)
+        await assertOkResponse(frisbeeRes)
+        // FIXME: can we pipe `frisbeeRes.body` directly to `res`?
+        // `frisbeeRes.body` is a Web API ReadableStream, `res` is a Node.js WritableStream
```
See https://nodejs.org/api/stream.html#streamreadabletowebstreamreadable

<img width="522" alt="Screenshot 2024-09-25 at 09 39 44" src="https://github.com/user-attachments/assets/73566bb1-6859-41d4-955a-55249b57881f">

You can either convert all to WebAPI streams, or all to Node.js streams, then continue as usual

bajtos replied:
Thank you for the pointer!

This seems to work:

```js
import * as stream from 'node:stream'
import { pipeline } from 'node:stream/promises'

// (...)

const frisbiiRes = await fetch(FRISBII_ADDRESS + req.url)
await assertOkResponse(frisbiiRes)
assert(frisbiiRes.body, 'frisbii response does not have any body')
await pipeline(stream.Readable.fromWeb(frisbiiRes.body), res)
```

juliangruber commented on indexer/lib/advertisement-walker.js:
```diff
@@ -151,7 +151,7 @@ export async function processNextAdvertisement ({
 
   try {
-    const { previousAdvertisementCid, ...entry } = await fetchAdvertisedPayload(
+    const { previousAdvertisementCid, entriesFetchError, ...entry } = await fetchAdvertisedPayload(
```
why is `entriesFetchError` plural and `entry` singular?

bajtos replied:
Let me explain how this works:

- First, we fetch the advertisement object. This object contains a CID linking to another object containing (payload block) entries.
- Second, we fetch the object with the entries.
- Finally, we take the first entry and discard the rest.

So, the error happens when we are fetching entries, but then we return only a single entry.

juliangruber commented on indexer/lib/advertisement-walker.js:
```diff
@@ -169,7 +169,10 @@ export async function processNextAdvertisement ({
+    if (entriesFetchError) {
+      state.entriesNotRetrievable = (state.entriesNotRetrievable ?? 0) + 1
+    }
+    const indexEntry = entry.pieceCid && entry.payloadCid ? entry : undefined
```
Do you mean

```suggestion
    const indexEntry = (entry.pieceCid && entry.payloadCid) ? entry : undefined
```

or

```suggestion
    const indexEntry = entry.pieceCid && (entry.payloadCid ? entry : undefined)
```

Could you please make this explicit?

juliangruber commented on indexer/lib/advertisement-walker.js:
```diff
@@ -252,23 +260,43 @@ export async function fetchAdvertisedPayload (providerAddress, advertisementCid,
+        err.statusCode,
+        /** @type {any} */(err).serverMessage ?? '<not found>'
+      )
+      return {
```
Why are we returning an object with `{ error }` instead of throwing it?

bajtos replied:
Yeah, this can be confusing, and I am thinking about redesigning my implementation.

What I wanted to accomplish in this PR:
- If the request for entries returns 404, we want to skip the advertisement and walk to the next advertisement in the chain. This is handled by returning `{ error }`
- If the request for entries fails with a different error, we want to retry it. This is handled by (re)throwing the error.

After thinking about this some more, I am leaning towards the following design:

- If the request for entries returns 404, we skip the advertisement. (No change from the current proposal.)
- If the request for entries fails differently, we retry it several times. If all attempts fail, then we skip the advertisement, too.

That way, we optimise for finishing the walk of all advertisements so that we can process new advertisements and avoid repeating the same failing request forever.

bajtos commented:
@juliangruber I reworked my PR as follows:

Inside `fetchAdvertisedPayload`:
- I added `pRetry` to retry when the server returns a 5xx error. (It does not make sense to retry after an error like 404 because the next request is almost certainly going to fail too.)
- I changed the return value to either contains `entry` object or an `error` string code.
- I recognise two errors - `CANNOT_FETCH_ENTRIES` and `CANNOT_DETERMINE_PIECE_CID`
- In particular, this function no longer throws when we cannot fetch the content for the Entries CID link.

Inside `processNextAdvertisement`, I added code to distinguish those two error codes and added a new property `state.adsMissingPieceCID`.

juliangruber commented on indexer/lib/advertisement-walker.js:
```diff
@@ -169,52 +170,68 @@ export async function processNextAdvertisement ({
-}} AdvertisedPayload */
+/**
+ * @param {unknown} err
+ * @param {string} providerAddress
```
```suggestion
 * @param {string} providerAddress
 * @returns {string}
```

bajtos replied:
TypeScript infers the return type.

<img width="740" alt="Screenshot 2024-10-02 at 13 34 08" src="https://github.com/user-attachments/assets/3343285e-6b40-45ad-9739-eef7f3dc4020">

But I can see how it can be helpful to see the return type in the JSDoc comment. I'll add `@returns {string | undefined}`

juliangruber commented on indexer/lib/advertisement-walker.js:
```diff
@@ -252,23 +269,53 @@ export async function fetchAdvertisedPayload (providerAddress, advertisementCid,
+  if (!pieceCid) {
+    debug('advertisement %s has no PieceCID in metadata: %j', advertisementCid, meta.deal)
+    return {
+      error: /** @type {const} */('CANNOT_DETERMINE_PIECE_CID'),
```
what is this type cast for?

bajtos replied:
Without this cast, the `error` would have the type `string` and auto-completion in VS Code won't work.

With the cast, the `error` has an enum type `'CANNOT_DETERMINE_PIECE_CID' | 'CANNOT_FETCH_ENTRIES'`.

Then, when you start writing a condition like `if (error === '`, VSCode can auto-complete the valid values.

bajtos commented on indexer/lib/advertisement-walker.js:
```diff
@@ -169,52 +170,68 @@ export async function processNextAdvertisement ({
     }
 
-    const indexEntry = entry.pieceCid ? entry : undefined
+    if (error === 'CANNOT_FETCH_ENTRIES') {
```
I'll use `ENTRIES_NOT_RETRIEVABLE`. We cannot retrieve the list of entries. We don't try to retrieve the entry itself.
